### ISSUE TITLE
IP range support for nodeippool settings (was: SSH connection lost )

### ISSUE THREAD

**Issue by Peter2121** [cbsd/cbsd#253]
FreeBSD version ( uname -a ): 
FreeBSD bimp-jails.xxxx.xx 11.1-RELEASE-p4 FreeBSD 11.1-RELEASE-p4 #0: Tue Nov 14 06:12:40 UTC 2017     root@amd64-builder.daemonology.net:/usr/obj/usr/src/sys/GENERIC  amd64

CBSD version ( cbsd version ): 
11.1.11

NAT is NOT configured, ipfw is NOT loaded in loader.conf, pf is NOT loaded in loader.conf.
Changing of sshd port was NOT asked during install.
sshd is listening on all addresses:
tcp4       0      0 *.22                   *.*                    LISTEN
One jail is created. 
When the jail is not started - ssh connections are possible from elsewhere. When the jail is started - only local ssh connections are possible.
tcpdump traces of remote ssh connection:
root@xxxxx:~ # tcpdump -ni vmx0 host 10.192.20.100
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on vmx0, link-type EN10MB (Ethernet), capture size 262144 bytes
18:00:06.546645 IP 10.192.20.100.23739 > 10.192.20.200.22: Flags [S], seq 319008763, win 65535, options [mss 1460,nop,wscale 6,sackOK,TS val 4002380513 ecr 0], length 0
18:00:09.559738 IP 10.192.20.100.23739 > 10.192.20.200.22: Flags [S], seq 319008763, win 65535, options [mss 1460,nop,wscale 6,sackOK,TS val 4002383527 ecr 0], length 0
18:00:12.765736 IP 10.192.20.100.23739 > 10.192.20.200.22: Flags [S], seq 319008763, win 65535, options [mss 1460,nop,wscale 6,sackOK,TS val 4002386733 ecr 0], length 0
18:00:16.039421 IP 10.192.20.100.23739 > 10.192.20.200.22: Flags [S], seq 319008763, win 65535, options [mss 1460,nop,wscale 6,sackOK,TS val 4002390007 ecr 0], length 0
18:00:19.320338 IP 10.192.20.100.23739 > 10.192.20.200.22: Flags [S], seq 319008763, win 65535, options [mss 1460,nop,wscale 6,sackOK,TS val 4002393288 ecr 0], length 0
18:00:22.596378 IP 10.192.20.100.23739 > 10.192.20.200.22: Flags [S], seq 319008763, win 65535, options [mss 1460,nop,wscale 6,sackOK,TS val 4002396564 ecr 0], length 0

**Comment by olevole** [comment]
Can you post additional info?

Result of:

1. 

```
cbsd jls display=jname,ip4_addr,interface
```

2.
When jail active:
```
jls -v
```

3.
Hoster network settings:
```
ifconfig
route -n get 0.0.0.0
```

**Comment by Peter2121** [comment]
```
# cbsd jls display=jname,ip4_addr,interface
JNAME  IP4_ADDR          INTERFACE
jab1   10.192.20.201/29  auto
```
Jail inactive:
```
# ifconfig
vmx0: flags=8843<UP,BROADCAST,RUNNING,SIMPLEX,MULTICAST> metric 0 mtu 1500
        options=60039b<RXCSUM,TXCSUM,VLAN_MTU,VLAN_HWTAGGING,VLAN_HWCSUM,TSO4,TSO6,RXCSUM_IPV6,TXCSUM_IPV6>
        ether 00:50:56:b0:16:2d
        hwaddr 00:50:56:b0:16:2d
        inet 10.192.20.200 netmask 0xffffff00 broadcast 10.192.20.255
        nd6 options=29<PERFORMNUD,IFDISABLED,AUTO_LINKLOCAL>
        media: Ethernet autoselect
        status: active
lo0: flags=8049<UP,LOOPBACK,RUNNING,MULTICAST> metric 0 mtu 16384
        options=600003<RXCSUM,TXCSUM,RXCSUM_IPV6,TXCSUM_IPV6>
        inet6 ::1 prefixlen 128
        inet6 fe80::1%lo0 prefixlen 64 scopeid 0x2
        inet 127.0.0.1 netmask 0xff000000
        nd6 options=21<PERFORMNUD,AUTO_LINKLOCAL>
        groups: lo
```
```
# route -n get 0.0.0.0
   route to: 0.0.0.0
destination: 0.0.0.0
       mask: 0.0.0.0
    gateway: 10.192.20.254
        fib: 0
  interface: vmx0
      flags: <UP,GATEWAY,DONE,STATIC>
 recvpipe  sendpipe  ssthresh  rtt,msec    mtu        weight    expire
       0         0         0         0      1500         1         0
```
Jail active:
```
# jls -v
   JID  Hostname                      Path
        Name                          State
        CPUSetID
        IP Address(es)
     7  xmpp.xxxxxx.xxx                  /cbsdjails/jails/jab1
        jab1                          ACTIVE
        2
        10.192.20.201
```
```
# ifconfig
vmx0: flags=8843<UP,BROADCAST,RUNNING,SIMPLEX,MULTICAST> metric 0 mtu 1500
        options=60039b<RXCSUM,TXCSUM,VLAN_MTU,VLAN_HWTAGGING,VLAN_HWCSUM,TSO4,TSO6,RXCSUM_IPV6,TXCSUM_IPV6>
        ether 00:50:56:b0:16:2d
        hwaddr 00:50:56:b0:16:2d
        inet 10.192.20.200 netmask 0xffffff00 broadcast 10.192.20.255
        inet 10.192.20.201 netmask 0xfffffff8 broadcast 10.192.20.207
        nd6 options=29<PERFORMNUD,IFDISABLED,AUTO_LINKLOCAL>
        media: Ethernet autoselect
        status: active
lo0: flags=8049<UP,LOOPBACK,RUNNING,MULTICAST> metric 0 mtu 16384
        options=600003<RXCSUM,TXCSUM,RXCSUM_IPV6,TXCSUM_IPV6>
        inet6 ::1 prefixlen 128
        inet6 fe80::1%lo0 prefixlen 64 scopeid 0x2
        inet 127.0.0.1 netmask 0xff000000
        nd6 options=21<PERFORMNUD,AUTO_LINKLOCAL>
        groups: lo
```

**Comment by olevole** [comment]
Looks like your jail settings overlap original network upon jail start, Since CBSD sets IP/MASK (as alias) according to your choices. You have 'C' network on hoster  ( 255.255.255.0 = /24 ) but jail setup with a small network ( /29 - netmask  10.192.20.200-10.192.20.207   (8) ).

I believe that this is the problem.
When you choose IP for jail in the same network as hoster, you need to inherit network settings, so valid value for your lab is 255.255.255.0, or /24. On the other hand you may IP without mask - /32

Try to change ip4_addr settings for your jail in

```
cbsd jconfig
```
or via 
```
cbsd jset
```
to:
~~~
10.192.20.201/24
~~~
or
~~~
10.192.20.201/32
~~~

And test again

**Comment by Peter2121** [comment]
OK, indeed, it works, thank you!!
But in this case the question is - how should I configure cbsd to attribute IP addresses to jails from an IP range of hoster's network? For example, I would like to have 201...210 addresses of hoster's network reserved for jails.
Globally, you explain in details the configuration of NAT etc. but the case of jails in the same network as hoster is not clearly documented (or I did not find this document).

**Comment by olevole** [comment]
Yeah, similar functionality would be very good and it's still in my todo.
The beginning of this implementation exists as a **nodeippool** settings in `cbsd initenv-tui', but at the moment it does not support range e.g from 201..210. At the moment you can specify C network in nodeippool, e.g:

10.192.20.0/24

and when ip4_addr='DHCP', CBSD will find the first IP starting from 10.192.20.{1}++ but I did not teach to use a mask. 

At the moment it's a dull stub with increment:

https://github.com/cbsd/cbsd/blob/master/tools/dhcpd#L50-L71

Therefore, you now need to manage IP addresses yourself for a while

**Comment by Peter2121** [comment]
So, as I understand:
- in the global configuration I should put the configuration IP of the hoster (10.192.20.0/24)
- every time I create a new jail, I should modify his IP to put it manually in the desired IP range

Is it true?

Anyway, it would be nice to add at least _starting_ address in the procedure you've mentioned, it does not seem to be too difficult (but this parameter needed to be saved in the global configuration).

**Comment by olevole** [comment]
> in the global configuration I should put the configuration IP of the hoster (10.192.20.0/24)

nodeippool is actual only when you use reserved 'DHCP' value in ip4_addr for jail. In this case, CBSD will find and set a free address in 10.192.20.{0-254} range.

However, since you need a narrower range it  does not fit ( nodeippool settings and 'DHCP'). Just set's valid/free IP from your range by hand:
```
cbsd jset ip4_addr=10.192.20.201/24
..
cbsd jset ip4_addr=10.192.20.210/24
```

> Anyway, it would be nice to add at least starting address in the procedure you've mentioned, it does not seem to be too difficult (but this parameter needed to be saved in the global configuration).

I agree with you, this is not a very complicated operation. I plan to do it in the foreseeable future.

Please leave this Issue open, as a reminder of this need ;-)

**Comment by olevole** [comment]
Fixed. Will be available in CBSD 11.1.16

**Comment by olevole** [comment]
commited to ports tree
